File: node/zkgroup/profiles/ProfileKey.ts
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import ByteArray from '../internal/ByteArray';
import NativeImpl from '../../NativeImpl';
import ProfileKeyCommitment from './ProfileKeyCommitment';
import ProfileKeyVersion from './ProfileKeyVersion';
import { UUIDType, fromUUID } from '../internal/UUIDUtil';

export default class ProfileKey extends ByteArray {
  static SIZE = 32;

  constructor(contents: Buffer) {
    super(contents, ProfileKey.checkLength(ProfileKey.SIZE));
  }

  getCommitment(uuid: UUIDType): ProfileKeyCommitment {
    return new ProfileKeyCommitment(
      NativeImpl.ProfileKey_GetCommitment(this.contents, fromUUID(uuid))
    );
  }

  getProfileKeyVersion(uuid: UUIDType): ProfileKeyVersion {
    return new ProfileKeyVersion(
      NativeImpl.ProfileKey_GetProfileKeyVersion(this.contents, fromUUID(uuid))
    );
  }
}

File: java/java/src/main/java/org/signal/zkgroup/profiles/ProfileKey.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup.profiles;

import java.util.UUID;
import org.signal.zkgroup.InvalidInputException;
import org.signal.zkgroup.VerificationFailedException;
import org.signal.zkgroup.internal.ByteArray;
import org.signal.client.internal.Native;

public final class ProfileKey extends ByteArray {

  public ProfileKey(byte[] contents) throws InvalidInputException {
    super(contents);
    try {
      Native.ProfileKey_CheckValidContents(contents);
    } catch (IllegalArgumentException e) {
      throw new InvalidInputException(e.getMessage());
    }
  }

  public ProfileKeyCommitment getCommitment(UUID uuid) {
    byte[] newContents = Native.ProfileKey_GetCommitment(contents, uuid);

    try {
      return new ProfileKeyCommitment(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ProfileKeyVersion getProfileKeyVersion(UUID uuid) {
    byte[] newContents = Native.ProfileKey_GetProfileKeyVersion(contents, uuid);

    try {
      return new ProfileKeyVersion(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}
